include/libtorrent/torrent_handle.hpp

```diff
@@ -276,6 +276,11 @@ namespace aux {
 		// guaranteed to finish in the same order as you initiated them.
 		void read_piece(piece_index_t piece) const;
 
+		// This function sets the starting piece (and optionally the ending piece)
+		// for the piece picker. It helps to start loading from the required piece
+		// in sequential mode. This only works in sequential mode. If no piece
+		// picker has been created, then the function does nothing.
```

Comment:
I think we've discussed this before, but from the use cases of Oleg and Denis, they wanted to facilitate seeking as well as reading media metadata (which may not be at the beginning) they didn't care much for an ending range, though they were okay with it, upon your suggestion or insistence. To keep the API surface to a minimum the present method with an optional end piece is my preference.

As far as whether the end should be open or closed, it is somewhat arbitrary, one is in conformance with, as you've mentioned, C++ iteration convention; the other matches human intuition. This is an API method, not a for loop, so my weak preference is for the latter. Though I can appreciate arguments for either way.

Reply:
> I think we've discussed this before, but from the use cases of Oleg and Denis, they wanted to facilitate seeking as well as reading media metadata (which may not be at the beginning) they didn't care much for an ending range, though they were okay with it, upon your suggestion or insistence. To keep the API surface to a minimum the present method with an optional end piece is my preference.

Yes, we have discussed this before, and every time (as I recall) we've stopped at this conclusion:

* there are only very special cases where it makes sense to set the sequential-end to the end of the torrent. Namely single file torrents, or where the sequential file being the last one in the torrent. In the general case, it does not make sense to sequential-download the next files in the torrent, as files are ordered arbitrarily.
* Sequential downloading more than necessary is hurting the swarm more than necessary. Hurting the swarm should not be the default.

I don't think it matters whether Oleg care about hurting the swarm or not. libtorrent's stance must represent the interest of all users, including the other peers on the swarm. If Oleg want to hurt the swarm more than necessary, that's fine, but I think it's reasonable to ask him to pass in one additional argument to do so deliberately.

> As far as whether the end should be open or closed, it is somewhat arbitrary, one is in conformance with, as you've mentioned, C++ iteration convention; the other matches human intuition. This is an API method, not a for loop, so my weak preference is for the latter. Though I can appreciate arguments for either way.

I think we may have different human intuitions. I would expect human intuition to include:

start + length = end

On the other hand, files (in v1 torrents) aren't necessarily aligned to pieces, so to find the first and last (or begin and end) pieces of a file isn't exactly trivial. It would probably be done similar to the [dump_torrent example](https://github.com/arvidn/libtorrent/blob/RC_2_0/examples/dump_torrent.cpp#L158).

Either way, it should be documented

Reply:
Let's not forget that Oleg has claimed that the changes to his patch have ruined it and made it bad at streaming, that he would therefore not use it. Furthermore he really cares about the 1.2 branch, not master.

I mean I know Oleg is hard to collaborate with - that is likely why he is by himself as a developer (the root cause of which might involve therapy sessions inquiring about his relationship with his father during his formative years) - but I don't want to make this about him. This is not personal. We should be professional. Let's take the emotion out of the decision.

If you may recall this whole issue started due to Denis not having an easy or performant way with the current API to seek within media or start playback. When it was suggested to also have an end range he was fine with it, though this was not his motivation or use case.

This is getting to be bikeshedding at this point, but having the default be the end of the torrent makes it easy for users, allowing them to fall into the "pit of success". That is the motivation here.

Reply:
I think you should strike (most of) your second paragraph, it's uncalled for.

> This is getting to be bikeshedding at this point, but having the default be the end of the torrent makes it easy for users, allowing them to fall into the "pit of success". That is the motivation here.

It makes it easy to do the wrong thing.
Do we agree that downloading sequentially contributes to poor swarm performance?
If so, I would also expect that we agree that it should be done to a minimum.

Reply:
Yes it can contribute to such. Though if you recall during our discussions on improving streaming, you've articulated numerous times that most of the time any adverse effects (or benefits) wouldn't matter as most torrents are well seeded anyway.

This API surface is sort of an esoteric facility; its use should be at a minimum, in any case. It would not perturb existing swarms.

Reply:
> Though if you recall during our discussions on improving streaming, you've articulated numerous times that most of the time any adverse effects (or benefits) wouldn't matter as most torrents are well seeded anyway.

Right, I believe that's the case. 

> This API surface is sort of an esoteric facility; its use should be at a minimum, in any case. It would not perturb existing swarms.

I don't see how that's an argument in favour of a default resulting in worse performance. Perhaps the "worse" is just a *little bit* worse in just a few cases, but the magnitude of how much worse it is doesn't seem very relevant.

Reply:
Arvid this default is what Denis and Oleg care about. Specifying the start of a torrent is the use case they articulated. While this present API - as you have made clear - is not fully positive on all metrics of interest, this is not a black and white thing, it is shades of gray. On the balance between making an easy and intuitive API for users and having an all performant interface that delivers maximally in all scenarios and corner cases, can't you agree that this default is a reasonable compromise?

Reply:
> On the balance between making an easy and intuitive API for users and having an all performant interface that delivers maximally in all scenarios and corner cases, can't you agree that this default is a reasonable compromise?

Perhaps if an argument can be made that this default actually delivers "an easy and intuitive API for users". I (currently) don't think it does. In fact, I think asking callers of this function to think about the end condition is a good thing.

You seem to suggest that passing in the end piece is a big hurdle. I don't see how it is, it's not *that* complicated to compute for the end of a file (and there are some utility functions available too).

Reply:
The API as it currently is *asks* callers of this method to think about the end condition. Your proposed solution *demands* they do so, when they don't wish to.

Ok Arvid, you suggested earlier in this thread that there can be an overload of a two parameter version of set_sequential_range(), would this compromise satisfy your demands?

Reply:
> The API as it currently is asks callers of this method to think about the end condition. Your proposed solution demands they do so, when they don't wish to.

Sure, yes. As far as I can tell, demanding the caller to think about it would be better than allowing them not to.

> Ok Arvid, you suggested earlier in this thread that there can be an overload of a two parameter version of set_sequential_range(), would this compromise satisfy your demands?

I think it would be reasonable if the default is the last piece of the file that `start` falls in. I think that default would be what 99% of users want.

It might be a bit weird if the piece has multiple files in it. I can see two ways of addressing that:

1. make a conservative assumption that the *last* file in the piece is the file that's being streamed
2. instead of taking `start` as a piece index, take a byte offset into the torrent. That would uniquely identify the file. (In this case the `end` probably also should be a byte offset)

Reply:
I think it could be done something like this:

```
// m_files is file_storage
// find the last file in piece `start`
auto const f = m_files.map_block(start, m_files.piece_size(start) - 1, 1).back().file_index;

// find the last piece in file `f`
auto const piece_size = m_files.piece_length();
auto const last = piece_index_t((m_files.file_offset(f) + m_files.file_size(f) + piece_size - 1) / piece_size);
```

Reply:
This is a lot of complexity for a proposed overload.

I will do a two parameter method without overload to satisfy your demands.

Reply:
> It helps to start loading from the required piece

I think you mean "it starts downloading from the specified piece".

> in sequential mode. This only works in sequential mode. If no piece

You say it only works in sequential mode, but this will enable sequential mode, right?

> picker has been created, then the function does nothing.

I think it's clearer to say "if the torrent does not have metadata" rather than "if no piece picker has been created". It's not documented anywhere when the piece picker is created, it's considered an implementation detail.

Also, I think you need to extend this documentation now to also specify what the single piece overload does. Namely, that it implicitly sets the end to the end of the file `first_piece` points to.